app/services/analytics.rb

```diff
@@ -4,7 +4,54 @@ class Analytics
   include AnalyticsEvents
   prepend Idv::AnalyticsEventsEnhancer
 
-  attr_reader :user, :request, :sp, :session, :ahoy
+  # Analytics middleware that sends the event to Ahoy
+  class AhoyMiddleware
+    attr_reader :ahoy
+
+    def initialize(ahoy: nil, request: nil)
+      @ahoy = ahoy || Ahoy::Tracker.new(request:)
+    end
+
+    def call(event)
+      ahoy.track(event[:name], event[:properties])
+      nil
+    end
+  end
+
+  # Analytics middleware that augments NewRelic APM trace with additional metadata.
+  class NewRelicMiddleware
+    def call(event)
+      # Tag NewRelic APM trace with a handful of useful metadata
+      # https://www.rubydoc.info/github/newrelic/rpm/NewRelic/Agent#add_custom_attributes-instance_method
+      ::NewRelic::Agent.add_custom_attributes(
+        user_id: event.dig(:properties, :user_id),
+        user_ip: event.dig(:properties, :user_ip),
+        service_provider: event.dig(:properties, :service_provider),
+        event_name: event[:name],
+        git_sha: IdentityConfig::GIT_SHA,
+      )
+
+      nil
+    end
+  end
+
+  class << self
+    # @return [Proc[]] The set of middleware Procs added to all new Analytics instances by default.
+    def default_middleware
+      @default_middleware ||= []
+    end
+
+    # @param [Proc[]] middlewares Middleware procs to add while block executes.
+    def with_default_middleware(*middlewares, &block)
```

Comment:
This is meant to support specs

Reply:
I would add that as a comment!
```suggestion
    # Added for use in specs only
    def with_default_middleware(*middlewares, &block)
```